### PR TITLE
Data store refactors (instances-ID, rules-ID)

### LINKED ISSUE #322: Make dataStore a real "database" like with custom adapter

### PR DESCRIPTION
**Pull request by tokland**
### :pushpin: References

* **Issue:** Closes #322

To run the migrations on the CLI: `$ yarn migrate 'http://admin:PASSWORD@localhost:8080'`

### :memo: Implementation

- Refactor syncRules setters (use generic update helpers).
- Creates in dataStore `instances-ID` (containing metadataMapping).
- Creates in dataStore `rules-ID` (containing builder+targetInstances).
- For the split: keep the class agnostic regardless on how data is persisted. Then define a type for the "main" element (in the list) and a "detail" type for the extended info. This split must taken in account only on DB operations (get, save and remove).
- Auto-load migration files from a folder (works for CRA/webpack and node/commonjs).
- Migrations are run in travis.yml. At some point, we should update the docker images.
- To run the migrations on the CLI: `$ yarn migrate 'http://admin:PASSWORD@localhost:8080'`
- When writing a migration, we must define the old/new types of the data structures used in that migration task. Note that we cannot rely on the current types in the app, as they may change in the future. For fields/object we don't care the type, use `unknown` (not `any`).

### CODE REVIEW

**Comment by EyeSeeTeaBotTest** [comment]
<h3>Tests failed with errors :rotating_light:</h3><br />Build commit Merge 46b4f44e0af607290a4b804d2a656a81dc475eae into 08d8b4e79b9af2d7248e52a777298a29fffb4d5d with commit id 73eb68e42c66e0c6c1fb473b733ad0b47c784068 on branch feature/data-store-refactors-322<br /><br />Travis log available at <a href=https://travis-ci.org/EyeSeeTea/metadata-synchronization/builds/664829199 target=_blank>builds/664829199</a><br /><br />See cypress results at <a href=https://eyeseetea.github.io/feedback-ci/EyeSeeTea/metadata-synchronization/pr/472/73eb68e42c66e0c6c1fb473b733ad0b47c784068 target=_blank>feedback-ci/EyeSeeTea/metadata-synchronization/pr/472/73eb68e42c66e0c6c1fb473b733ad0b47c784068</a>

File: src/models/instance.ts [hunk 1]
```diff
@@ -148,7 +162,14 @@ export default class Instance {
         const element = exists ? instance.data : { ...instance.data, id: generateUid() };
 
         if (exists) await instance.remove(api);
-        return saveData(api, instancesDataStoreKey, element);
+
+        const detailsKey = Instance.getDetailsKey(instance);
+        const detailsData: InstanceDetailsData = { metadataMapping: instance.metadataMapping };
+        const response: Response = await saveDataStore(api, detailsKey, detailsData)
+            .then(() => ({ status: true }))
+            .catch(error => ({ status: false, error }));
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/models/instance.ts]: There are quite a lot of helpers with signature `Promise<void> + exception` to  Promise<Response>`. I've created a `toResponse` to abstract this transformation.

File: src/migrations/index.ts [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+const migrations: Migration[] = [
+    { version: 1, fn: initial, name: "Initial" },
+    { version: 2, fn: instancesById, name: "Create instances-ID" },
+    { version: 3, fn: rulesById, name: "Create rules-ID" },
+];
+
+export const currentVersion =
+    _(migrations)
+        .map(info => info.version)
+        .max() || 0;
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/migrations/index.ts]: The first expression it's still a migration. Something like: `_.maxBy(migrations, migration => migration.version)?.version ?? 0`. It's a bit convoluted...

File: src/migrations/index.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        return { newVersion, migrationsToApply };
+    }
+}
+
+if (require.main === module) {
+    const [baseUrl] = process.argv.slice(2);
+    if (!baseUrl) throw new Error("Usage: config.ts DHIS2_URL");
+    const runner = new MigrationRunner({ migrations, baseUrl, debug: console.debug });
+    runner.migrate();
+}
```
[the remark this answers is not in the document]
Comment (tokland) [reply src/migrations/index.ts]: Yes, the React component was still not pushed (it is now). But I wanted to keep this, having a runnable script is nice, and it's very easy to provide a CLI interface like this.

**Comment by SferaDev** [comment]
All the changes LGTM! Thanks @tokland

**Comment by tokland** [comment]
> All the changes LGTM! Thanks @tokland

Thank you for the suggestions! Still missing the axios retries, wildo.

**Comment by tokland** [comment]
Cool! 

Ok, I'll wait for @adrianq review to do all the visual tweaks together.

**Comment by SferaDev** [comment]
I just thought of a scenario we might want to address before/after merging this PR.

What if the instance data store version is bigger than the application one? (ie. outdated app)

Would we want to show a message that the app requires an update and please contact an administrator?

**Comment by tokland** [comment]
> What if the instance data store version is bigger than the application one? (ie. outdated app) 
> Would we want to show a message that the app requires an update and please contact an administrator?

Good point. I think so, we should not allow the user to use the app.

**Comment by tokland** [comment]
@SferaDev Done! b0861877

**Comment by adrianq** [comment]
It works like a charm. Regarding the layout, maybe we can show messages in a reverse way (ala analytics report in DHIS2). Merging this anyway as it is not a big deal

**Pull request merged**